### ISSUE TITLE
opentype.js having problem processing ttf font extracted from ttc

### ISSUE THREAD

**Issue by bitinn** [nodebox/opentype.js#139]
OK, I will admit this is a confusing case, but this is what we believe we know so far:

- OS X El Capitan comes with a new default system font for Chinese users, aka `PingFang`. So I am trying to analyze it with opentype.js' Font Inspector.
- Between their Developer Preview Beta1 and Beta2, Apple made some changes to this font, naturally, I want to see what's the difference.
- The steps I use is as following:

1. get the font from `/System/Library/Fonts/`
2. unpack with ttc2ttf https://gist.github.com/kayahr/2479682
3. load using http://nodebox.github.io/opentype.js/font-inspector.html

Problem is: while the resulting ttfs from DP1 works, result from DP2 font doesn't, font inspector gives us `TypeError: Cannot read property 'glyphIndexMap' of null` which I believe is an error from `opentype.js`, not just the Font Inspector script.

Initially we believe this is a problem with `ttc2ttf`, because DP2 ttc is using some kind of offsets to the same ttf 2 different names, ie. both DP1 and DP2 `PingFang.ttc` contains only 12 actual font files, but DP2 adds extra alias to each fonts, so 24 names show up.

But using `fonttools` and Apple Font Book we are not able to find any faults with these files.

So I have a theory, somewhere in `opentype.js`, there is a bug that's only triggered under this scenario. Would you take a look when available.

I will share the DP2 PingFang.ttc here in case you need it (probably against its font license, but hopefully for a good cause)

https://www.dropbox.com/s/qnfy6q41zkji330/PingFang_DP2.ttc.zip?dl=0

For an introduction on PingFang, see my blog post: http://bitinn.net/11280/

**Comment by blackgear** [comment]
PingFang DP2 removed "platform 3" cmap tables:
see DP1 ttx file:
```
  <cmap>
    <tableVersion version="0"/>
    <cmap_format_4 platformID="0" platEncID="3" language="0">
    <cmap_format_12 platformID="0" platEncID="4" format="12" reserved="0" length="189520" language="0" nGroups="15792">
```
DP2 ttx file:
```
  <cmap>
    <tableVersion version="0"/>
    <cmap_format_4 platformID="0" platEncID="3" language="0">
    <cmap_format_12 platformID="0" platEncID="4" format="12" reserved="0" length="185584" language="0" nGroups="15464">
    <cmap_format_4 platformID="3" platEncID="1" language="0">
    <cmap_format_12 platformID="3" platEncID="10" format="12" reserved="0" length="185584" language="0" nGroups="15464">
  </cmap>
```

see: https://github.com/nodebox/opentype.js/blob/master/dist/opentype.js#L2720 
```
    // The cmap table can contain many sub-tables, each with their own format.
    // We're only interested in a "platform 3" table. This is a Windows format.
    cmap.numTables = parse.getUShort(data, start + 2);
    var offset = -1;
    for (i = 0; i < cmap.numTables; i += 1) {
        var platformId = parse.getUShort(data, start + 4 + (i * 8));
        var encodingId = parse.getUShort(data, start + 4 + (i * 8) + 2);
        if (platformId === 3 && (encodingId === 1 || encodingId === 0)) {
            offset = parse.getULong(data, start + 4 + (i * 8) + 4);
            break;
        }
    }
```
without cmap table with platformId == 3, the code is broken as designed.

well, actually the platformId == 0 and platformId == 3 use the same table data in DP1:
![untitled](https://cloud.githubusercontent.com/assets/4780199/8350951/3cccf520-1b5c-11e5-9a4a-5b630c56490a.jpg)

The offset of this to kind of table(see the offset value in Subtable) are just the same.

The table with platformId == 0 is equal to the platformId == 3.

Change the `        if (platformId === 3 && (encodingId === 1 || encodingId === 0)) {` to `        if (platformId === 0 && (encodingId === 1 || encodingId === 0)) {` can perfect fix the problem.

But remember this situation is ONLY for PingFang, and there is no guarantee for all fonts,

**Comment by bitinn** [comment]
Thx @blackgear!

https://github.com/nodebox/opentype.js/blob/master/src/tables/cmap.js#L19-L36

Looks like #105 is related to this problem.

**Comment by fdb** [comment]
@bitinn not quite. PlatformID = 0 is used for the ["Unicode platform"](https://www.microsoft.com/typography/otspec/name.htm). Bug #105 refers to PlatformID = 1 (Mac), EncodingId = 0 (ancient 256-character encoding).

It seems that Apple's encoding is correct. Platform ID = 0, Encoding ID = 3 is Unicode 2.0 and onwards semantics. This is similar to Platform ID = 3, Encoding ID = 1.

I might try this so that PlatformID 3 and PlatformID 0 can use the same cmap parsers. Let's see what the results are with a number of different fonts.

**Comment by HinTak** [comment]
Yes, Apple uses Platform ID = 0, Encoding ID = 3, and that to most intents and purposes the same as Platform ID = 3, Encoding ID = 1 . They are certainly equivalent for the common fonts supporting European languages.

**Comment by Jolg42** [comment]
cmap 12 read support was just added with PR #207 so that will help I think!
By the way @blackgear what software did you used in your screenshot to open & parse the .ttf?

**Comment by blackgear** [comment]
@Jolg42 "Synalyze It! - much more than a Hex Editor for Mac OS X"

**Comment by Jolg42** [comment]
@blackgear Thank you 😉 It works well! I will update the [awesome-typography](https://github.com/Jolg42/awesome-typography) repo asap!
